jax/experimental/jax2tf/jax2tf.py

```diff
@@ -103,67 +103,79 @@ def add_unit(v: TfValOrUnit, aval: core.AbstractValue):
-      out = fun(*args, **kwargs)
-      return out, grad_disabled
-
-    return inner_wrapper(*tensor_args)
-
-  return wrapper
-
-
-def convert(fun):
+def convert(fun,
+            with_gradient=False):
   """Transforms `fun` to be executed by TensorFlow.
 
   Args:
     fun: Function to be transformed. Its arguments and return value should be
       JAX arrays, or (nested) standard Python containers (tuple/list/dict)
       thereof.
+    with_gradient: if set, will add a tf.custom_gradient to the converted
+      function, by converting the ``jax.vjp(fun)``. Only first-order
+      differentiation is supported for now. If the converted function is
+      saved in a SavedModel, the custom gradients are currently lost and
+      an error will be raised if a gradient computation is attempted.
 
   Returns:
     A version of `fun` that expects TfVals as arguments (or
     tuple/lists/dicts) thereof, and returns TfVals as outputs.
   """
   api._check_callable(fun)
 
-  @disable_gradient
-  def wrapped_fun(*args: TfValOrUnit) -> TfValOrUnit:
-    # TODO(necula): remove the jit disabling once we handle all control-flow.
-    # Disabling the jit helps to avoid some unsupported jax primitives.
-    # E.g. scan will be statically unrolled.
-    f = lu.wrap_init(fun)
+  def converted_fun(*args: TfVal) -> TfVal:
+    # This function may take pytrees of TfVals. We can only set
+    # tf.custom_gradient on functions that take a flat argument list.
     args_flat, in_tree = tree_util.tree_flatten((args, {}))
     for a in args_flat:
       if not _is_tfvalorunit(a):
         msg = (f"Argument {a} of type {type(a)} of jax2tf.convert(f) should "
                "be NumPy array, scalar, tf.Variable, or tf.Tensor")
         raise TypeError(msg)
-    flat_fun, out_tree = flatten_fun(f, in_tree)
-    out_flat = _interpret_fun(flat_fun, args_flat)
-    return tree_util.tree_unflatten(out_tree(), out_flat)
 
-  return wrapped_fun
+    f = lu.wrap_init(fun)
+    # out_tree_thunk() will be the output tree, after running _interpret_fun.
+    flat_fun, out_tree_thunk = flatten_fun(f, in_tree)
+
+    # Prepare the grad_fn for tf.custom_gradient.
+    def converted_grad_fn(*out_cts_flat: TfVal, **kwargs):
+      # TODO(cl/318778369): change **kwargs with variables=None
+      variables = kwargs.get("variables", [])
+      if variables:
+        raise ValueError("Unexpected variables used in forward pass. "
+                         "This should not happen for first-order differentiation. "
+                         f"variables={variables}")
+
+      def fun_vjp_jax(args_jax, out_cts_jax):
+        # One may think that we can get the pullback while we are converting
+        # the main function in the first place. That is problematic, because the
+        # pullback may contain captured tracers from the conversion of the
+        # main function. Those tracers will confuse the conversion of the
+        # pullback. So, we construct the vjp anew.
```

Comment:
I'm not sure I fully understand the implications of this, but does this mean if I have a custom gradient that closes over some expensive to compute intermediate value from forward and uses it as part of computing the gradient that when running via jax2tf we will actually always recompute this value (because the custom gradient for our function is "disconnected" from the forward pass)?

As a concrete example for this pseudocode:

```
@custom_gradient
def op(x):
  intermediate = x ** 2
  def grad(g):
    return 5 * g * intermediate
  return intermediate ** 2
```

I think we would produce a tf graph with two copies of the forward pass, one to compute the forward result and one in order for us to compute grad (aka. we'll recompute `intermediate` in order to compute `grad`). A smart enough compiler could fix this for us I guess. I'm a bit worried because this is a pretty common use case for custom gradients so we should at least call it out as a major limitation.

Reply:
You are right about this performance issue. It is in fact a bit worse, in the sense that you pay this cost even in the absence of custom gradients. This is likely fixable, but will require somewhat deeper messing with tracers and linear_util. I plan to look at it later, for now I wanted to get the main pieces in place.